Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -200,59 +205,98 @@ proof_wanted Language.isContextFree_iff {L : Language T} :
-    {w : List T} (hgw : w ∈ g.reverse.language) :
-    w.reverse ∈ g.language := by
-  convert g.reverse_derives hgw
-  simp [List.map_reverse]
-
-lemma ContextFreeGrammar.mem_reverse_language_iff_reverse_mem_language (g : ContextFreeGrammar T)
-    (w : List T) :
-    w ∈ g.reverse.language ↔ w.reverse ∈ g.language := by
-  refine ⟨reverse_mem_language_of_mem_reverse_language _, fun hw => ?_⟩
-  rw [← ContextFreeGrammar.reverse_involutive g] at hw
-  rw [← List.reverse_reverse w]
-  exact g.reverse.reverse_mem_language_of_mem_reverse_language hw
+@[simp] lemma reverse_reverse (g : ContextFreeGrammar T) : g.reverse.reverse = g := by
+  simp [reverse, Finset.map_map]
+
+lemma reverse_involutive : Involutive (reverse : ContextFreeGrammar T → ContextFreeGrammar T) :=
+  reverse_reverse
+
+lemma reverse_bijective : Bijective (reverse : ContextFreeGrammar T → ContextFreeGrammar T) :=
+  reverse_involutive.bijective
+
+lemma reverse_injective : Injective (reverse : ContextFreeGrammar T → ContextFreeGrammar T) :=
+  reverse_involutive.injective
+
+lemma reverse_surjective : Surjective (reverse : ContextFreeGrammar T → ContextFreeGrammar T) :=
+  reverse_involutive.surjective
+
+lemma produces_reverse : g.reverse.Produces u.reverse v.reverse ↔ g.Produces u v :=
+  (Equiv.ofBijective _ ContextFreeRule.reverse_bijective).exists_congr
+    (by simp [ContextFreeRule.reverse_involutive.eq_iff])
+
+alias ⟨_, Produces.reverse⟩ := produces_reverse
+
+@[simp] lemma produces_reverse_comm : g.reverse.Produces u v ↔ g.Produces u.reverse v.reverse :=
+  (Equiv.ofBijective _ ContextFreeRule.reverse_bijective).exists_congr
+    (by simp [ContextFreeRule.reverse_involutive.eq_iff])
+
+protected lemma Derives.reverse (hg : g.Derives u v) : g.reverse.Derives u.reverse v.reverse := by
+  induction hg with
+  | refl => rfl
+  | tail _ orig ih => exact ih.trans_produces orig.reverse
+
+lemma derives_reverse : g.reverse.Derives u.reverse v.reverse ↔ g.Derives u v :=
+  ⟨fun h ↦ by convert h.reverse <;> simp, .reverse⟩
+
+@[simp] lemma derivees_reverse_comm : g.reverse.Derives u v ↔ g.Derives u.reverse v.reverse := by
+  rw [iff_comm, ← derives_reverse, List.reverse_reverse, List.reverse_reverse]
+
+lemma generates_reverse : g.reverse.Generates u.reverse ↔ g.Generates u := by simp [Generates]
+
+alias ⟨_, Generates.reverse⟩ := generates_reverse
+
+@[simp] lemma generates_reverse_comm : g.reverse.Generates u ↔ g.Generates u.reverse := by
+  simp [Generates]
+
+@[simp] lemma language_reverse : g.reverse.language = g.language.reverse := by ext; simp
+
+end ContextFreeGrammar
 
 /-- The class of context-free languages is closed under reversal. -/
 theorem Language.IsContextFree.reverse (L : Language T) :
-    L.IsContextFree → L.reverse.IsContextFree :=
-  fun ⟨g, hg⟩ => ⟨g.reverse, hg ▸ Set.ext g.mem_reverse_language_iff_reverse_mem_language⟩
+    L.IsContextFree → L.reverse.IsContextFree := by
+  rintro ⟨g, rfl⟩; exact ⟨g.reverse, by simp⟩
```

Comment:
Can we remove the `;` and make it two lines please? Everything else in this file is written "one tactic a line".

Reply:
https://github.com/leanprover-community/mathlib4/blob/32419692d138999119817d8612c02c254d26fcba/Mathlib/Computability/ContextFreeGrammar.lean#L80

would like to disagree!

Reply:
Oooops. It looks like a refactor somebody did after me.

Reply:
Apparently it comes straight from #6385 ! Maybe past Martin would disapprove current Martin 😆